security/src/main/java/uk/ac/ox/ndph/mts/security/authorisation/AuthorisationService.java

```diff
@@ -121,13 +121,15 @@ public boolean authorise(String requiredPermission, List<String> entitiesSiteIds
                 return false;
             }
 
-            if (!hasValidPermissions(requiredPermission, roleAssignments)) {
+            var rolesAssignmentsWithPermission = getRolesAssignmentsWithPermission(requiredPermission, roleAssignments);
+
+            if (rolesAssignmentsWithPermission.isEmpty()) {
                 return false;
             }
 
             List<SiteDTO> sites = siteServiceClient.getAllSites();
```

Comment:
Just getting my head around the logic and the rest calls here - do I have this right?
(1) get all role assignments for practitioner P   (from practitioner-service)
(2) get permissions for those roles (from role-service) 
(3) filter role assignments to remove those which connect to roles which don't connect to permission X 
(4) get all sites (from site-service) and buildsite tree hierarchy
(5) filter role assignments again, this time by site S and child-sites CS.

I wondered whether we could do the site filtering at an earlier stage, e.g. by asking practitioner-service only for role assignments at site S, for practitioner P.  But I don't think that can work unless practitioner-service knows the site-tree hierarchy, otherwise it will miss child sites is that right?   

I wondered about performance here with so many rest calls.  I think the logic is necessary to achieve the story.  So my only thought is - maybe cache some of the data?

Reply:
You are right , you can't because of the child hierarchy. Regarding caching, we thought about it during the design too. You can cache the whole site tree. I'll check if there is already a story for that.